src/modules/Game.class.js

```diff
@@ -1,68 +1,483 @@
 'use strict';
 
-/**
- * This class represents the game.
- * Now it has a basic structure, that is needed for testing.
- * Feel free to add more props and methods if needed.
- */
 class Game {
-  /**
-   * Creates a new game instance.
-   *
-   * @param {number[][]} initialState
-   * The initial state of the board.
-   * @default
-   * [[0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0],
-   *  [0, 0, 0, 0]]
-   *
-   * If passed, the board will be initialized with the provided
-   * initial state.
-   */
-  constructor(initialState) {
-    // eslint-disable-next-line no-console
-    console.log(initialState);
-  }
-
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  static Status = {
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(
+    initialState = Array.from({ length: 4 }, () => Array(4).fill(0)),
+  ) {
+    this.state = [...initialState];
+    this.status = Game.Status.idle;
+    this.score = 0;
+    this.initialState = [...initialState];
+    this.undoStack = []; // Stack to store previous game states
+  }
+
+  getScore() {
+    return this.score;
+  }
+
+  getState() {
+    return this.state;
+  }
+
+  getStatus() {
+    this.updateStatus();
+
+    return this.status;
+  }
+
+  start() {
+    this.status = Game.Status.playing;
+    this.state = this.initialState.concat();
+    this.createRandomTile();
+    this.createRandomTile();
+    this.printTiles();
+  }
+
+  restart() {
+    this.clearTheBoard();
+    this.printTiles();
+    this.status = Game.Status.idle;
+    this.score = 0;
+    this.undoStack = []; // Clear undo stack on restart
```

Comment:
There is a small bug here, steps to reproduce:

1. Start the game. 
2. Make a few moves to receive some score. 
3. Click on a 'Restart' button. Note the 'score' value.
4. Click on a 'Start' button. Note the 'score' value. 
5. Make a move. Note a 'score' value. 

Expected behaviour: 
the score equals to 0 as soon as a user clicked on a 'restart' button. 

Current behaviour:
the score is reseted only after step 5.